### ISSUE TITLE
`/messages` takes ages to back-paginate through private history visibility

### ISSUE THREAD

**Issue by richvdh** [matrix-org/synapse#13128]
If you join a room with `joined` history visibility, then your client will attempt to backfill history. Synapse does the following:

* loads a batch of events from the db
* filters out the ones you can't see (ie, all of them)
* returns the remainder

So the result looks like this:

```json
{
  "chunk": [],
  "start": "t23104-3060236712_757284974_3318886_1437115222_1456937831_3359724_521254283_4632180264_0",
  "end": "t23085-659955750_757284974_3318886_1437115222_1456937831_3359724_521254283_4632180264_0"
}
```

If the history is long, this can go on for ages (like, hours) before we reach the actual start of the room.

**Comment by DMRobertson** [comment]
What's the fix here? Something like

> If the room history is `joined`, only query fpr events with $THING greater than that the $THING of your most recent not-join->join event

for some value of $THING? (stream id? origin server ts?)

**Comment by richvdh** [comment]
> What's the fix here? Something like
> 
> > If the room history is `joined`, only query fpr events with $THING greater than that the $THING of your most recent not-join->join event

No, because it's possible that room history viz has not always been `joined`, or that you were not previously a member of the room - in either case you should be able to see earlier chunks of history.

But some sort of caching of which sections of timeline have relaxed history viz, combined with more intelligent traversal of the timeline based on the user's membership events, seems possible (if thorny).

It gets even more thorny when you throw federation into the mix.
